### PR TITLE
feat(bash-debug-adapter): add bash-debug-adapter

### PR DESCRIPTION
**Pull request by Issafalcon**
Tested locally using WSL on Ubuntu 20.04

- Installs the bash debug adapter with no errors
- Not entirely sure if I'm missing anything as I copied code from other examples of debug adapter install features

### CODE REVIEW

**Comment by Issafalcon** [comment]
Hi. Just thought I'd add bash debug adapter to the mix as it was missing. Let me know if anything else needs adding, or if you'd prefer I also raise an issue and link it to this PR

**Comment by Issafalcon** [comment]
OK. So I just realised that trying to unzip the release file doesn't work quite as expected due to the release of some untagged versions of the debugger (i.e. https://github.com/rogalmic/vscode-bash-debug/releases/download/untagged-438733f35feb8659d939/bash-debug-0.3.9.vsix). So mason is looking for the vsix file using the wrong URL with the current code. I think my one successful download might have been a red herring. 

I'm not sure if installing using npm compile script (which is how I've installed it in the past) is viable for this, or how to fix the unusually formed URL that mason attempt to download from. In addition, the vscode-bash-debug adapter hasn't been updated for a long while, which makes me wonder if this is still the best way to debug shell scripts in neovim!

**Comment by williamboman** [comment]
> OK. So I just realised that trying to unzip the release file doesn't work quite as expected due to the release of some untagged versions of the debugger (i.e. https://github.com/rogalmic/vscode-bash-debug/releases/download/untagged-438733f35feb8659d939/bash-debug-0.3.9.vsix). So mason is looking for the vsix file using the wrong URL with the current code. I think my one successful download might have been a red herring.

Ah yeah that's annoying. In this case we'll have to interface directly with the GitHub client to release data and extract the first suitable asset file (like test against `%.vsix$`). This would have to be done while respecting `ctx.requested_version` (if empty the latest release should be used). I'll see if I can add something to the PR

**Comment by williamboman** [comment]
@williambotman /apply

```diff
diff --git a/lua/mason-core/managers/github/init.lua b/lua/mason-core/managers/github/init.lua
index 55f3600..a0336d3 100644
--- a/lua/mason-core/managers/github/init.lua
+++ b/lua/mason-core/managers/github/init.lua
@@ -88,7 +88,7 @@ end
 ---@param processor async fun(opts: table)
 local function release_file_processor(filename, processor)
     ---@async
-    ---@param opts {repo: string, asset_file: string|fun(release: string):string}
+    ---@param opts {repo: string, version: Optional|nil, asset_file: string|fun(release: string):string}
     return function(opts)
         local release_file_source = M.release_file(opts)
         std.download_file(release_file_source.download_url, filename)
diff --git a/lua/mason-registry/bash-debug-adapter/init.lua b/lua/mason-registry/bash-debug-adapter/init.lua
index 26f91f3..4277f2e 100644
--- a/lua/mason-registry/bash-debug-adapter/init.lua
+++ b/lua/mason-registry/bash-debug-adapter/init.lua
@@ -1,7 +1,9 @@
 local Pkg = require "mason-core.package"
 local github = require "mason-core.managers.github"
+local github_client = require "mason-core.managers.github.client"
 local _ = require "mason-core.functional"
 local path = require "mason-core.path"
+local Optional = require "mason-core.optional"
 
 return Pkg.new {
     name = "bash-debug-adapter",
@@ -12,16 +14,32 @@ return Pkg.new {
     ---@async
     ---@param ctx InstallContext
     install = function(ctx)
+        local repo = "rogalmic/vscode-bash-debug"
+        ---@type GitHubRelease
+        local release = ctx.requested_version
+            :map(function(version)
+                return github_client.fetch_release(repo, version)
+            end)
+            :or_else_get(function()
+                return github_client.fetch_latest_release(repo)
+            end)
+            :get_or_throw()
+
+        ---@type GitHubReleaseAsset
+        local release_asset = _.find_first(_.prop_satisfies(_.matches "%.vsix$", "name"), release.assets)
+
         github
             .unzip_release_file({
-                repo = "rogalmic/vscode-bash-debug",
-                asset_file = _.compose(_.format "bash-debug-%s.vsix", _.gsub("^v", "")),
+                repo = repo,
+                asset_file = release_asset.name,
+                version = Optional.of(release.tag_name),
             })
             .with_receipt()
+
         ctx.fs:rmrf(path.concat { "extension", "images" })
         ctx:link_bin(
             "bash-debug-adapter",
-            ctx:write_node_exec_wrapper("bash-debug-adapter", path.concat { "out", "bashDebug.js" })
+            ctx:write_node_exec_wrapper("bash-debug-adapter", path.concat { "extension", "out", "bashDebug.js" })
         )
     end,
 }

```

**Comment by williamboman** [comment]
@williambotman /fixup

File: lua/mason-registry/bash-debug-adapter/init.lua [hunk 1]
```diff
@@ -0,0 +1,45 @@
+local Pkg = require "mason-core.package"
+local github = require "mason-core.managers.github"
+local github_client = require "mason-core.managers.github.client"
+local _ = require "mason-core.functional"
+local path = require "mason-core.path"
+local Optional = require "mason-core.optional"
+
+return Pkg.new {
+    name = "bash-debug-adapter",
+    desc = [[Debug your bash scripts.]],
```
Comment (williamboman) [inline lua/mason-registry/bash-debug-adapter/init.lua:10]:
```suggestion
    desc = [[Bash shell debugger, based on bashdb.]],
```

**Comment by Issafalcon** [comment]
Nice @williamboman ! Out of interest, the 'languages' list on the package, and the subsequent language mappings that are added are for typescript and JavaScript. Am I correct in assuming that these mapping refer to the languages that the package itself was written in?

**Pull request merged**